### ISSUE TITLE
proposal: combine get and list endpoints

### ISSUE THREAD

**Issue by krancour** [akuity/kargo#1553]
There's little difference between get and list requests.

* In the CLI, there is only get. It's "get all" (no name specified == list) or get (a) specified resource(s). `kubectl` works the same way.

* Our CLI already only ever uses the list endpoints -- even for a gets.
    * If specific resources were requested, they're selected _client-side_ from the complete list -- [something I have already called out as being inefficient](https://github.com/akuity/kargo/issues/1098).

What I propose is, for all resource types, combining the API endpoints for get and list -- I would probably opt for using "get" as the name for the combined endpoints:

* Takes a project name and an _optional_ list of resource names.
* All cases return list of resources _and_ list of resource names _not found_:
    * If no resource names are specified, returns entire list of resources.
    * If names are specified:
        * Selects those that were requested _server side_.
        * Returns them in the order they were requested.
        * Returns names of resource not found, also in the order they were requested

This issue would supersede #1098

**Comment by krancour** [comment]
I am going to implement new endpoints for #1523 in this manner so that there is a concrete example of this in case we choose to apply it more broadly.

**Comment by jessesuen** [comment]
I don't think we should do this. Get and list are inherently different privileges and we are trying to follow kubernetes RBAC model. So someone may have permissions to get a single stage but not all.

Also, with a unified endpoint for get and list, we lose out on REST-like properties like returning 404 if the resource by name doesn't exist. With the combined API, we would return 200 but with an empty list, and expect clients to deduce from response body about the presence of the object (instead of return codes).

Clients would additionally need to check length of the returned lists (to prevent index out of bounds) to ensure > 0, whereas if they used a traditional GET API, they can blindly access the object as long as they don't get a 404. 

In short, I think dealing with 404s is more ergonomic than dealing with 0, 1, 2+ length lists without response code to indicate the result.

**Comment by krancour** [comment]
@jessesuen all good points. I have solutions for the RBAC issues and some of the ergonomic issues, but rather than dive into those, I'll take a step back to explain my motivations:

> Our CLI already only ever uses the list endpoints -- even for a gets.

    This undermines RBAC already. You'd _need_ list perms in order to get. No good.

> If specific resources were requested, they're selected client-side from the complete list

    Return the world when you wanted one or two things? This is wildly inefficient. See #1098

Both of these already need fixing. The simple fix is to make all get commands use the existing get endpoints instead of using list endpoints, but that creates a _new_ problem -- namely that if you `kargo get stages foo bar bat baz`, we now need to make _four_ API calls instead of one. What if it's a get of 50 resources?

This got me thinking that get endpoints could pretty easily be modified to receive a _list_ of resource names as an argument instead of just a single resource name and _return_ a list as well. If you then say an empty list of resource names means _list all_, you can see where one the get endpoint can handle all three cases of get one, get multiple, and list all and a dedicated list endpoint is no longer needed.

So those were my motivations...

As I write this, I'm coming to the conclusion that I am over-optimizing for the unlikely case of "getting" a _large_ number of resources by name. If you agree that it is indeed an over-optimization, then I am happy to scrap this idea.
